src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -264,17 +264,22 @@ public bool TryCancel()
                 return true;
             }
 
-            public void Dispatch()
+            public void Dispatch(bool processAsyncOperationSynchronously)
             {
                 ManualResetEventSlim? e = Event;
                 if (e != null)
                 {
                     // Sync operation.  Signal waiting thread to continue processing.
                     e.Set();
                 }
+                else if (processAsyncOperationSynchronously)
```

Comment:
I think we need to change sightly how this is structured, in particular for the e != null case. We get there if a synchronous operation is performed on a Socket on which an asynchronous operation was ever performed. In such cases, we've permanently moved the socket to be non-blocking, which means we need to simulate the blocking behavior of all subsequent sync operations, and we do that by using a MRES instead of a callback. We don't want to require a thread pool thread just to set that event, as doing so could lead to thread pool starvation, with a sync operation waiting on a thread pool thread for a work item to be processed by the thread pool that will unblock it. So, we want the epoll thread to set such an MRES rather than queuing a work item to do it.

Reply:
> So, we want the epoll thread to set such an MRES rather than queuing a work item to do it.

I am afraid that it would require us to make the enqueueing more complex and decrease throughput in a noticeable way.

To set MRES on a epoll thread we would need to check two queues (send and receives) which would require to take two locks:

https://github.com/dotnet/runtime/blob/6ef538d46fc971f351a22cbe854717f4ada59d89/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L832

and then perform 0 to two casts (depending on whether queues are empty or not):

https://github.com/dotnet/runtime/blob/723e7f25fd651506dc18792c53cef3ea622b10f9/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L137

Ofc we would have to do that for every epoll event returned by epoll wait.

With @kouvel proposal after we receive an epoll notification we just add a bunch of simple events to a queue (this is very fast) and schedule a work item to the thread pool.

Reply:
> I am afraid that it would require us to make the enqueueing more complex and decrease throughput in a noticeable way.

We only need to do it _if_ the event is for a sync operation, which can be checked cheaply.

Reply:
> We only need to do it if the event is for a sync operation, which can be checked cheaply.

But we don't know this when we are receiving the epoll notification. To check this we need to translate the socket handle to socket context:

https://github.com/dotnet/runtime/blob/6ef538d46fc971f351a22cbe854717f4ada59d89/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs#L333

and then take the two locks that I've described above.

Reply:
Then we could consider having a dedicated epoll thread for sync-over-async.

I understand your pushback, but this is a big deal.

Reply:
A way to address this could be to enqueue the `AsyncOperation` to the `ConcurrentQueue` here (instead of events from the epoll). Would that work? I may be overlooking something.

Reply:
> Convince me it's not if you disagree :

I agree that it's an important problem that we should solve ;)

> Then we could consider having a dedicated epoll thread for sync-over-nonblocking sockets

I was thinking about that too. I am not sure if the sync|async information is available when we are adding given socket to epoll. Moving it from one epoll to another would be possible, but it would require some synchronization.

Another idea I had was to set given socket to blocking before performing the synchronous operation (afaik currently we always set reads and writes to non-blocking).

 https://github.com/dotnet/runtime/blob/6ef538d46fc971f351a22cbe854717f4ada59d89/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs#L833

But I don't know if it can be set more than one time (so user could do sync write and later async for example).

@tmds do you know? Also, Tom do you have any ideas for addressing this problem?

Reply:
> Another idea I had was to set given socket to blocking before performing the synchronous operation (afaik currently we always set reads and writes to non-blocking).

Socket starts life as Blocking (i.e. its nonblocking mode is false).  Sends and Receives performed on it are truly synchronous down to the OS.  The first time you perform an async operation on the Socket, we flip nonblocking to be true.  After that point, any synchronous Sends/Receives you perform on it are emulated to be synchronous via blocking waiting for the epoll to set the MRES.  We avoid flipping nonblocking back to false due to various race conditions.

Reply:
Maybe it can be recorded on SocketAsyncContext if there are any pending sync reads or writes, or at least if the first operation that would be processed is a sync operation, we'd still have to look up the context from the concurrent dictionary on the epoll thread but it may be able to do a cheap check to see if it should set the event on the epoll thread.

Reply:
> A way to address this could be to enqueue the AsyncOperation to the ConcurrentQueue here (instead of events from the epoll). Would that work? I may be overlooking something.

I think this should work. Though we shouldn't enqueue the operation itself, but execution of the queue. The operation may be cancelled by time this gets dequeued for execution.

Reply:
> The operation may be cancelled by time this gets dequeued for execution.

Looks like `ProcessQueuedOperation` accounts for that.
So the suggested change is something like:
* `processAsyncOperationSynchronously` becomes `processAsyncOperationOnConcurrentQueue`, and enqueues to the `ConcurrentQueue<AsyncOperation>`
* `HandleEvents` gets called directly on the epoll thread, `ConcurrentQueue` processing is deferred to the ThreadPool after calling `HandleEvents`.

Reply:
Taking the lock for each operation queue on the epoll thread seems to reduce RPS by ~20-25 K. In the new commits I made changes to track whether the first operation in each queue is synchronous and to check it speculatively from the epoll thread. The speculative check along with the concurrent dictionary dequeue is not making a noticeable difference to RPS. Considering that the alternative is not incorrect, it seems like a speculative check would be enough to avoid the starvation issue.

I don't think there's a need to queue `AsyncOperation` to the queue, but anyway it involves taking the lock, and doing so seems to reduce RPS by an additional 15-20 K RPS on top of taking the locks, not sure fully why, though there is a bit more work involved in extracting those.

Reply:
The code would be simpler using `ConcurrentQueue<AsyncOperation>`.

> Taking the lock for each operation queue on the epoll thread seems to reduce RPS by ~20-25 K.

I plan to take a shot at replacing these locks with `Interlocked` operations.

Reply:
> it involves taking the lock, and doing so seems to reduce RPS by an additional 15-20 K RPS on top of taking the locks, not sure fully why, though there is a bit more work involved in extracting those.

What lock does this refer to?

Reply:
It's the lock taken in `SocketAsyncContext.HandleEvent` before this change (linked above by @adamsitnik), in this change the lock taken in `SocketAsyncContext.ProcessSyncEventOrGetAsyncEvent`.

> The code would be simpler using `ConcurrentQueue<AsyncOperation>`

Agreed, I wanted to get that to work and tried it first but there currently seem to be obstacles to doing that.